docs/root/overview.md

```diff
@@ -0,0 +1,189 @@
+# Nighthawk: architecture and key concepts
+
+## High level interaction model
+
+**Process** creates one or more **Workers**. **Worker** will run **Sequencer**,
+which in turn queries **RateLimiter** for [request-release
+timings](terminology.md#request-release-timings). When it is time to release a
+request, **BenchmarkClient** will be requested to do so by **Sequencer**.
+**BenchMarkClient** will then ask its underlying **Pool** to create a
+**StreamDecoder** release the actual request. **StreamDecoder** will query the
+request data it needs to send from the configured **RequestSource**, and send it
+off. **StreamDecoder** will emit events as it progresses (pool ready,
+completion, etc), and timings will subsequently be recorded into **Statistic**
+as well as get bubbled up to **Sequencer** for tracking in-flight work and
+**Statistic** bookkeeping.
+
+**Sequencer** will query the configured **TerminationPredicates** to terminate
+when and how to terminate execution. When all **Workers** have finished,
+**Process** will collect the results from all **Workers** via
+**OutputCollector**, and run **OutputFormatter** to transform to the requested
+output format.
+
+However, it is said that pictures say more then words, so here is a block
+diagram: (TODO(#249): create a real diagram)
+
+![Draft diagram](draft-high-level-diagram.png)
+
+## Notable upcoming changes
+
+Calling out two new concepts that may get proposed in the future, and cause some
+churn in the code base as we inject them.
+
+### Phases
+
+One notable addition / change that may get proposed in the near future is the
+introduction of **Phase**. **Phase** would represent a distinct stage of an
+execution, like, for example, warm-up. It would then be useful to have
```

Comment:
(nit) Feels like this is bit too many commas in a row, how about:

...a distinct stage of an execution, for example a warm-up.